crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -179,17 +196,27 @@ impl Paginator {
             });
         }
 
+        let reset_state_guard =
+            ResetStateGuard { target: PaginatorState::Initial, state: self.state.clone() };
+
         // TODO: do we want to lazy load members?
         let lazy_load_members = true;
 
         let response =
             self.room.event_with_context(event_id, lazy_load_members, num_events).await?;
 
+        // NOTE: it's super important to not have any `await` after this point, since we
+        // don't want the task to be interrupted anymore, or the internal state
+        // may become incorrect.
+
         let has_prev = response.prev_batch_token.is_some();
         let has_next = response.next_batch_token.is_some();
-        *self.prev_batch_token.lock().await = response.prev_batch_token;
-        *self.next_batch_token.lock().await = response.next_batch_token;
+        *self.prev_batch_token.lock().unwrap() = response.prev_batch_token;
+        *self.next_batch_token.lock().unwrap() = response.next_batch_token;
 
+        // Forget the reset state guard, so its Drop method is not called.
+        mem::forget(reset_state_guard);
```

Comment:
Like in the other comment, I don't think we should use `forget()` here. Or at all in the crate for that matter.